btree/btree.c

```diff
@@ -1501,6 +1502,253 @@ static void generic_move(struct nd *src, struct nd *tgt,
 	node_fix(tgt, tx);
 }
 
+
+/**
+ *  -------------------------------------------------------
+ *  Section START - Credit calculation for BTree Operations
+ *  -------------------------------------------------------
+ */
+
+//In case of fix key, fix value node format: max_ksize, max_vsize will be fix key, value size else estimate max sizes
+m0_bcount_t max_ksize; 
+m0_bcount_t max_vsize;
+m0_bcount_t max_node_size;
+m0_bcount_t internal_vsize; //for internal node, value size is fix
+
+/** threshold can be used in btree_move_key_value_credit() to identify 
+ * what fraction of node needs to be move in case of overflow/underflow
+ */
+enum threshold{
+	MAX_MOVE = 7,
+	MIN_MOVE = 5,
+	ALL_MOVE = 10
+};
+
+enum {
+	BTREE_ALLOC_SHIFT = 0,
+};
+
+static void btree_mem_alloc_credit(const struct m0_btree  *btree,
+				   m0_bcount_t             size,
+				   struct m0_be_tx_credit *accum)
+{
+	m0_be_allocator_credit(NULL, M0_BAO_ALLOC_ALIGNED, size,
+			       BTREE_ALLOC_SHIFT, accum);
+}
+
+static void btree_node_alloc_credit(const struct m0_btree  *tree,
+				    struct m0_be_tx_credit *accum)
+{
+	btree_mem_alloc_credit(tree, max_node_size, accum);
+}
+
+/**
+ * Calculates the credit needed to update @nr nodes and adds this credit to
+ * @accum.
+ */
+static void btree_node_update_credit(struct m0_be_tx_credit *accum,
+				     m0_bcount_t             nr)
+{
+	struct m0_be_tx_credit cred = {};
+
+	/* @todo: why multiplication with 2 is needed? */
+	m0_be_tx_credit_mac(&cred,
+			    &M0_BE_TX_CREDIT_TYPE(max_node_size), 2);
```

Comment:
I think this is wrong.
1. For a btree, its node size should be known at this moment. We need to take a "btree" as parameter and then get its node size.
2. if the node size is unknown, we may use the `max_node_size`. But, the use of M0_BE_TX_CREDIT_TYPE() here is wrong. This is only to measure the size of a "m0_bcount_t". This is apparently wrong.